### PR TITLE
devicetree: replace DT_HAS_DRV_INST with DT_INST_FOREACH

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
Make drivers multi-instance wherever possible using DT_INST_FOREACH.
This allows removing DT_HAS_DRV_INST in favor of making drivers just
do the right thing regardless of how many instances there are.

There are a few exceptions:

- some SoC drivers with ifdefs inside code for each enabled instance.
  These could be converted to IF_ENABLED() or COND_CODE_1() eventually
  to use DT_INST_FOREACH. For the sake of expediency, just inline the
  DT_HAS_DRV_INST expansion for now.

- SoC drivers which are explicitly single-instance. Again for the sake
  of expediency, drop a BUILD_ASSERT in those cases to make sure the
  assumption that all supported SoCs have at most one available
  instance is valid, failing fast otherwise.

### CODE REVIEW

File: drivers/sensor/nxp_kinetis_temp/temp_kinetis.c [hunk 1]
```diff
@@ -180,7 +180,13 @@ static int temp_kinetis_init(struct device *dev)
 	return 0;
 }
 
-#if DT_HAS_DRV_INST(0)
+/* Driver is single-instance */
+#define CHECK_INSTANCE_NUMBER(inst) \
+	BUILD_ASSERT((inst) == 0, "unsupported temp instance");
```
Comment (galak) [inline drivers/sensor/nxp_kinetis_temp/temp_kinetis.c:185]: Should this check DT_NUM_INST instead.

Comment (mbolivar-nordic) [reply drivers/sensor/nxp_kinetis_temp/temp_kinetis.c:185]: Good idea

**Comment by mbolivar-nordic** [comment]
I've fixed the designware DMA driver bugs discussed on slack. According to @galak , @tbursztyka  agrees the devices weren't being instantiated correctly. I assume that means that it's safe to fix them and use DT_INST_FOREACH.

**Comment by mbolivar-nordic** [comment]
```
WARNING:TRAILING_SEMICOLON: macros should not use a trailing semicolon
```

Do we care? I don't personally, but I can use the workaround people have been doing where they end the instantiation macro in a function instead.

**Review by galak — APPROVED** [review]
LGTM.  We should probably open a issue to cleanup uses of DT_HAS_NODE_STATUS_OKAY(DT_DRV_INST(x)).

**Comment by galak** [comment]
@henrikbrixandersen curious about the kinetis temperature driver only supporting a single instance, but nxp_ke1xf.dtsi having 3.  Is that correct?

**Comment by galak** [comment]
> ```
> WARNING:TRAILING_SEMICOLON: macros should not use a trailing semicolon
> ```
> 
> 
> Do we care? I don't personally, but I can use the workaround people have been doing where they end the instantiation macro in a function instead.

Does it break if you remove the `;`.  At least for the ones that call DEVICE_AND_API_INIT() you should be able to safely remove it.

**Review by dcpleung — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
> At least for the ones that call DEVICE_AND_API_INIT() you should be able to safely remove it.

DEVICE_AND_API_INIT needs a trailing semicolon

**Comment by mbolivar-nordic** [comment]
> LGTM. We should probably open a issue to cleanup uses of DT_HAS_NODE_STATUS_OKAY(DT_DRV_INST(x)).

Will do after this gets merged.

**Comment by henrikbrixandersen** [comment]
> @henrikbrixandersen curious about the kinetis temperature driver only supporting a single instance, but nxp_ke1xf.dtsi having 3. Is that correct?

@galak Essentially, they all access the same, internal temperature sensor. The temperature can be read using either of the available ADCs, so all ADCs have their own (disabled) temperature sensor instance. It is up to the user of the temperature sensor to decide, which one to use (have the application share the ADC with the temperature sensor to reduce code size, use a dedicated ADC for reading temperature to avoid interfering with application code, ...).

I am okay with changing the driver to allow for multiple, enabled instances although from a user point of view it doesn't make much sense.

**Comment by mbolivar-nordic** [comment]
> I am okay with changing the driver to allow for multiple, enabled instances although from a user point of view it doesn't make much sense.

Please let me know if there's something nonsensical here and I'll put it back.

**Comment by galak** [comment]
> @galak Essentially, they all access the same, internal temperature sensor. The temperature can be read using either of the available ADCs, so all ADCs have their own (disabled) temperature sensor instance. It is up to the user of the temperature sensor to decide, which one to use (have the application share the ADC with the temperature sensor to reduce code size, use a dedicated ADC for reading temperature to avoid interfering with application code, ...).
> 
> I am okay with changing the driver to allow for multiple, enabled instances although from a user point of view it doesn't make much sense.

mostly curiosity on my part.

**Review by galak — APPROVED** [review]

**Comment by galak** [comment]
Ok w/the checkpatch errors, going to merge w/override.

**Pull request merged**